xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,201 @@
+	clusterUpdate xdsclient.ClusterUpdate
+
+	// This boolean determines whether this Node has received an update or not. This isn't the best practice,
+	// but this will protect a list of Cluster Updates from being constructed if a cluster in the tree has not received
+	// an update yet.
+	receivedUpdate bool
+
+	clusterHandler *clusterHandler
+
+}
+
+// CreateClusterNode creates a cluster node from a given clusterName. This will also start the watch for that cluster.
+func createClusterNode(clusterName string, xdsClient xdsClientInterface, topLevelHandler *clusterHandler) *clusterNode {
+	print("|createClusterNode() called on ", clusterName, "|")
+	c := &clusterNode{
+		clusterHandler: topLevelHandler,
+	}
+	print("About to communicate with the xds client for:", clusterName, "|")
+	// Communicate with the xds client here.
+	c.cancelFunc = xdsClient.WatchCluster(clusterName, c.handleResp)
+	return c
+}
+
+// This function cancels the cluster watch on the cluster and all of it's children.
+func (c *clusterNode) delete() {
+	c.cancelFunc()
+	for _, child := range c.children {
+		child.delete()
+	}
+}
+
+// Construct cluster update (potentially a list of ClusterUpdates) for a node.
+func (c *clusterNode) constructClusterUpdate() ([]xdsclient.ClusterUpdate, error) { // THIS LOGIC IS WRONG
+	print("Tries constructing a cluster update for Cluster ", c.clusterUpdate.ServiceName, ". ")
+	// If the cluster has not yet received an update, the cluster update is not yet ready.
+	if !c.receivedUpdate {
+		return nil, errors.New("Tried to construct a cluster update on a cluster that has not received an update.")
+	}
+
+	// Base case - LogicalDNS or EDS. Both of these cluster types will be tied to a single ClusterUpdate.
+	if c.clusterUpdate.ClusterType != xdsclient.ClusterTypeAggregate {
+		print("Returning the cluster update for service (base node):", c.clusterUpdate.ServiceName)
+		return []xdsclient.ClusterUpdate{c.clusterUpdate}, nil
+	}
+
+	// If an aggregate construct a list by recursively calling down to all of it's children.
+	var childrenUpdates []xdsclient.ClusterUpdate
+	for _, child := range c.children {
+		childUpdateList, err := child.constructClusterUpdate()
+		if err != nil {
+			return nil, err
+		}
+		childrenUpdates = append(childrenUpdates, childUpdateList...)
+	}
+	print("Returning the update list for an aggregate cluster")
+	return childrenUpdates, nil
+}
+
+// handleResp handles a xds response for a particular cluster. This function also
+// handles any logic with regards to any child state that may have changed.
+// Problem statement: perhaps this logic is wrong
+func (c *clusterNode) handleResp(clusterUpdate xdsclient.ClusterUpdate, err error) {
+	c.clusterHandler.clusterMutex.Lock()
+	defer c.clusterHandler.clusterMutex.Unlock()
+	print("SERVICE NAME:", clusterUpdate.ServiceName)
+	if err != nil { // Write this error for run() to pick up in CDS LB policy.
+		c.clusterHandler.updateChannel<-clusterHandlerUpdate{chu: nil, err: err}
```

Comment:
This send may block if there's an item in the channel.

Reply:
Good catch. Changed.

Reply:
Added an opportunistic drain before this error send.